provider/blockdev_mirror_base.py

```diff
@@ -0,0 +1,169 @@
+        self.clone_vm.create()
+        self.clone_vm.verify_alive()
+
+        self.env.register_vm("%s_clone" % self.clone_vm.name, self.clone_vm)
+
+    def add_target_data_disks(self):
+        """Hot plug target disks to VM with qmp monitor"""
+        for tag in self._target_images:
+            disk = self.target_disk_define_by_params(
+                self.params.object_params(tag), tag)
+            disk.hotplug(self.main_vm)
+            self.trash.append(disk)
+
+    def blockdev_mirror_serial(self):
+        """Run block-mirror one by one"""
+        try:
+            for idx, source_node in enumerate(self._source_nodes):
+                backup_utils.blockdev_mirror(self.main_vm, source_node,
+                                             self._target_nodes[idx],
+                                             **self._backup_options[idx])
+        finally:
+            memory.drop_caches()
+
+    def blockdev_mirror_parallel(self):
+        """Run block-mirror and other operations in parallel"""
+        # parallel_tests includes function names separated by space
+        # e.g. parallel_tests = 'stress_test', we should define stress_test
+        # function with no argument
+        parallel_tests = self.params.objects("parallel_tests")
+        targets = list([getattr(self, t)
+                        for t in parallel_tests if hasattr(self, t)])
+
+        # block-mirror on all source nodes is in parallel too
+        for idx, source_node in enumerate(self._source_nodes):
+            targets.append(
+                partial(backup_utils.blockdev_mirror, vm=self.main_vm,
+                        source=source_node, target=self._target_nodes[idx],
+                        **self._backup_options[idx])
+            )
+
+        try:
+            utils_misc.parallel(targets)
+        finally:
+            memory.drop_caches()
+
+    def _check_mirrored_block_node_attached(self, source_qdev, target_node):
+        out = self.main_vm.monitor.query("block")
+        for item in out:
+            if (item["qdev"] == source_qdev
+                    and item["inserted"].get("node-name") == target_node):
+                break
+        else:
+            self.test.fail("Device(%s) is not attached to target node(%s)"
+                           % (source_qdev, target_node))
+
+    def check_mirrored_block_nodes_attached(self):
+        """All source devices attach to the mirrored nodes"""
+        for idx, target in enumerate(self._target_nodes):
+            self._check_mirrored_block_node_attached(
+                self._source_images[idx], target)
+
+    def blockdev_mirror(self):
+        """Need to be implemented in specific test case"""
+        raise NotImplementedError
+
+    def do_test(self):
+        self.blockdev_mirror()
+        self.check_mirrored_block_nodes_attached()
+        self.clone_vm_with_mirrored_images()
+        self.verify_data_files()
+        self.remove_files_from_system_image()
```

Comment:
Hi Yongxue, do_test was implemented here because most block-mirror test cases follow the same test steps, just like what do_test does here, for others, they almost follow the same steps, with just a little difference, e.g. create a new file after block-mirror, so if we implement do_test in specific test case, we have to duplicate many code lines...

The difference between these test cases is the block_mirror, I have implemented three bock_mirror_* functions, which can be called in specific test cases.

This class likes a virtual c++ abstract class, it cannot be used to define an BlockdevMirrorBaseTest object, it can only be derived by children --- the specific test cases.

Reply:
Okay, I got your idea, looks good to me.